PitchShiftEffect: extend effect options

davidchocholaty opened the pull request:
This PR extends the settings options of the Pitch shift effect. The implemented options are Range Mode, Semitones Mode and Wide Range option.

The Range Mode changes the range of the Pitch knob. The neutral mode (knob in the default position) keeps the range from the lowest pitch to the highest pitch. The positive mode (knob in the maximum position) keeps the range from the default pitch to the highest pitch. The negative mode (knob in the minimum position) keeps the range from the default pitch to the lowest pitch. Every mode works correctly with set the Wide Range option too.

The Wide Range option doubles the default Pitch knob scale. So, when the default scale of the pitch knob is from -1 octave to +1 octave, with set the option on the range is from -2 octaves to +2 octaves.

Lastly, the semitones mode changes the Pitch knob scale from the continuous mode to the semitones mode which works in the chromatic scale. By default, this option is on.

Another option, that will be implemented, will be the Formant option which works with the same-named option of the RubberBand library.

---

**TODO**:
- [x] Implementation of the Formant option.

davidchocholaty committed:
PitchShiftEffect: add wide range option

This commit adds the Wide Range option for the Pitch shift effect.
This option doubles the scale of the pitch knob, which means,
that with the toggle on, the knob range is not in -1 octave
and +1 octave, but in doubled range, so in range -2 octaves
and +2 octaves. Based on the changes, the knob now works
for both situations in the exponential scale of the power of 2.

davidchocholaty committed:
PitchShiftEffect: add semitones mode option

This commit adds the Semitones option for the Pitch shift effect.
This option switches the pitch scale into the semitones mode
in a musical approach known. So, turning the Pitch knob is moved
along the chromatic scale.

davidchocholaty committed:
math: add sgn function

This commit adds the function for calculating the signum function.
This function returns the sign of the value. For example,
for integer values, it means 1 for positive values, 0 for zero values
and -1 for negative values.

davidchocholaty committed:
PitchShiftEffect: add range mode option

This commit adds the Range option for the Pitch shift effect.
This option changes the range of the pitch knob. There are three
possible options based on the knob position. The neutral mode (knob
in the default position) keeps the range from the lowest pitch
to the highest pitch. The positive mode (knob in the maximum position)
keeps the range from the default pitch to the highest pitch.
The negative mode (knob in the minimum position) keeps the range
from the default pitch to the lowest pitch. Every mode works correctly
with set the Wide Range option too.

davidchocholaty committed:
PitchShiftEffect: add a formant preserving option

This commit adds the Formant Preserving option for the Pitch shift
effect. This option changes the settings of formant processing
using the RubberBand library.  The default option doesn't use
any special formant processing. When the option is turned on,
the handling of the formant shape is activated and the spectral envelope
of the unshifted signal is preserved.

davidchocholaty committed:
PitchShiftEffect: remove insignificant TODO

This commit removes the forgotten TODO for the Range Mode option
description.

Swiftb0y commented on src/util/math.h:
```diff
@@ -79,3 +79,9 @@ inline const T db2ratio(const T a) {
     return static_cast<T>(pow(10, a / 20));
 }
+
+template<typename T>
+inline const T sgn(const T a) {
+    static_assert(std::is_arithmetic_v<T>, "the value has to be an arithmetic type");
+    return (a > T(0)) - (a < T(0));
+}
```
A few suggestions. To explain them: templated functions are always inline, so we can omit it. Use c++20 concepts instead of the static_assert since the result looks nicer and probably results in a nicer error message. When T is unsigned, `a < T(0)` will always be false, so compilers will warn you about it, so I put a `if constexpr` there. I also added some static casts since the implicit casts might break and/or cause compiler warnings and errors. Also made `constexpr` because why not.
```suggestion
/// https://en.wikipedia.org/wiki/Sign_function
template<typename T>
requires std::is_arithmetic_v<T>
constexpr T sgn(const T a) {
    // silence -Wtype-limits
    if constexpr (std::is_unsigned_v<T>) {
        return static_cast<T>(a > T(0))
    } else {
        return static_cast<T>(a > T(0)) - static_cast<T>(a < T(0));
    }
}
```

davidchocholaty replied:
Okay, thank you very much for the detailed explanation. Your version looks good to me. It is really a cleaner solution.

Swiftb0y commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +71,69 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
     pitch->setNeutralPointOnScale(0.0);
     pitch->setRange(-1.0, 0.0, 1.0);
 
+    EffectManifestParameterPointer rangeMode = pManifest->addParameter();
+    rangeMode->setId("rangeMode");
+    rangeMode->setName(QObject::tr("Mode"));
+    rangeMode->setShortName(QObject::tr("Mode"));
+    rangeMode->setDescription(QObject::tr(
```
IMO this is knob is unnecessary. What was your motivation for introducing it?

davidchocholaty replied:
The idea is based on the demo of the Pioneer DJM-900NX2 build-in effects including the Pitch Shift effect. As is shown in the [demo video](https://youtu.be/9BTKCT7NHIM?t=964), with turning the "TIME" knob, using the effect "LEVEL/DEPTH" knob, the Pitch goes only up or down based on the mentioned "TIME" knob setting (IMO it is not to best understandable solution, but don't care). Based on that idea, I think that it can be useful to work with the pitch in only one way (up or down) and that working with the knob in his limit bounds (minimum bound and maximum bound) is much easier. You don't have to take care to go from the middle knob position to maximum and back (and not go under the middle position). With that, the Pitch knob is much more sensitive, because the range, which was previously only the half of range (from the middle position to maximum for pitching up) is now recalculated to the whole knob range. In summary, it looks cool to me to allow working with the effect as same as on the Pioneer mixer (and maybe some other brand's mixers too), but that is just my point of view.

I agree, that using the knob for only 3 possible settings is not the clearest solution, but it cannot be allowed more modes simultaneously.

daschuer replied:
The DJM-900NX2 TIME knob behaves differently. It is our original Pitch knob with unity at center. It sets the range of the effect when using the (additional) "LEVEL/DEPTH" knob.
So instead of introducing "Mode" and "wideRange" you may just introduce a "Range" knob, that set the min/max of the Pitch knob. 

This maps to the Pioneer knobs like this: 
* TIME = Pitch 
* "LEVEL/DEPTH" = Range

daschuer replied:
Yes, but the range knob has no regions, it is continuously. That's why Phil Harris eagerly turns it to be sure to hit the max range. The manual says it has a Rage from -50 ... +100%  whatever that means. Translated into speed it would be -12 ... +12 semitones. https://docs.pioneerdj.com/Manuals/DJM_900NXS2_DRI1300A_manual/?_ga=2.239935050.831276606.1638182063-764262911.1599494699&page=16
Does that make sense? I think it is much more when listening to the video. 
The proposed -24 to + 24 sound reasonable.

davidchocholaty replied:
Perfect, thank you so much. I think I understand it completely now.

Swiftb0y commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +71,69 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
+    semitonesMode->setName(QObject::tr("Semitones"));
+    semitonesMode->setShortName(QObject::tr("Semitones"));
+    semitonesMode->setDescription(QObject::tr(
+            "Set the pitch scale to the semitones mode "
+            "instead of the contiguous mode (set as default)."));
```
I don't think these explanations are very intuitive. I have some proposals which seem more user friendly. wdyt?
```suggestion
            "Change pitch in semitone-steps instead of smoothly));
```

davidchocholaty replied:
Agree, looks good. Thank you.

Swiftb0y commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +71,69 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
+    formant->setName(QObject::tr("Formant"));
+    formant->setShortName(QObject::tr("Formant"));
+    formant->setDescription(QObject::tr(
+            "Applies the Formant Preserving processing "
+            "and activates the handling of the formant shape."));
```
```suggestion
            "Preserve formants"));
```

davidchocholaty committed:
math: clean up sgn function

This commit cleans the sgn function which implements the sign function
as widely known. The static assert for checking, if a type
is an arithmetic, is replaced using the required keyword
for the same situation. Newly introduced is checking if the type
is unsigned. If yes, the calculation can be simplified. Implicit casts
are replaced with explicit casts.

davidchocholaty committed:
PitchShiftEffect: semitones mode description

This commit replaces the Semitones Mode toggle description
with a more intuitive version.

davidchocholaty committed:
PitchShiftEffect: replace formant description

This commit replaces the Formant toggle description
with a more intuitive version.

davidchocholaty committed:
PitchShiftEffect: remove range mode and wide range

This commit removes the Range Mode option and Wide Range option
for the Pitch shift effect. These options should be replaced with
only one option which will handle the Range of the Pitch knob.

davidchocholaty committed:
Pitch shift effect: add Range option

This commit adds the Range knob for the Pitch shift effect.
The Range option changes the range of the Pitch knob. The Range knob
works linearly and the Pitch knob range is from 0 to 2 octaves.
The default Range knob position is in the middle, which means
that the range of the Pitch knob is from -1 octave to +1 octave.

daschuer commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +69,34 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
+    formant->setName(QObject::tr("Formant"));
+    formant->setShortName(QObject::tr("Formant"));
+    formant->setDescription(QObject::tr(
+            "Preserve formants."));
```
```suggestion
            "Preserve formants, the resonance frequencies of human vocal tract and other instruments."));
```

I must admit I was not clear what exactly formats are when starting play with pitch shifting. 
Maybe this one liner is enough to give at least a hint, for others like me. 

Or we may pick a sentence that is more catchy but less exact. 

"Preserve formants to compensate chipmunk or growling voices."

Swiftb0y replied:
```suggestion
            "Preserve the resonant frequencies of human vocal tract and other instruments (formants)"));
```
Thats better imo. Easy explanation upfront and then the technical term afterwards for those that are familiar with it.

daschuer replied:
Pick what works for you, or an entire different string. As long we explain the user a bit more what this is about it should be fine.

daschuer commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +69,34 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
+
+    EffectManifestParameterPointer formant = pManifest->addParameter();
+    formant->setId("formant");
+    formant->setName(QObject::tr("Formant"));
```
It looks like RB "engine fine" actually supports to shift the formants. If we consider this, the name "Formant" is better used for the formant rotary knob. This one is "Formant Preserving" in case we use the LV2 variant. Is the length still OK?

davidchocholaty replied:
Yeah, in the R3 version the formant can be shifted using a knob. With introducing the R3 version in the Pitch shift effect, this option of course could be added and if it would be, the identical toggle name will be the problem too. On the other hand, the "Formant Preserving" name is too wide, see the print screen:

![image](https://user-images.githubusercontent.com/84876095/185974426-f9350bde-0a6c-4242-84ee-adc313cc24d3.png)

It looks like the maximum number of characters is about 10.

Swiftb0y replied:
Then just name it formant and put the detailed explanation in the tooltip imo.

daschuer replied:
Ok, but at least the Id should not be "formant". Since the name can be changed at any time, but not the ID.

davidchocholaty replied:
Okay, that looks good to me.

Swiftb0y commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +69,34 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
+    semitonesMode->setName(QObject::tr("Semitones"));
+    semitonesMode->setShortName(QObject::tr("Semitones"));
+    semitonesMode->setDescription(QObject::tr(
+            "Change pitch in semitone-steps instead of smoothly."));
```
I purposefully suggested "smooth" instead of "continuous" because these texts should be as easy to understand as possible. IMO knowing the word "continuous" requires a higher level of english skills than "smooth". Even though "continuous" would probably be the better antonym.

davidchocholaty replied:
> Sounds more natural for me, not sure though ...
You could have even a smooth change to full semitone steps, which is a different feature.

I agree, based on some grammar upgrades, the `"Change the pitch in semitone steps instead of continuously."` version looks good to me.

davidchocholaty replied:
@Swiftb0y sorry, based my browser did not update your answer, I see it until now. IMHO the "continuously" looks more accurately in some way to me. To go a little bit deeper in the English point of view, both words belong to the B2 level, but it's true that the "continuously" word is rather for more technical English users. Of course, I am open-minded to both versions.

daschuer replied:
Since this is a translated text, there is no much need to be understandable for non English speaker. As long it is correct English it works for me. 

It just jumps to my eyes initially because It look unusual to me. But I can*t really judge. Maybe a native speaker can jump in, @ywwg? This is not a blocking issue, though.

davidchocholaty committed:
PitchShiftEffect: update semitones description

The commit updates the description of the semitones mode.
To be more concrete, the word "smoothly" is replaced with "continuously"
and some grammar updates are added too.

davidchocholaty committed:
PitchShiftEffect: extend formant description

The commit extends the description of the Formant option
with a more exact version and with an easy-to-understand hint.

davidchocholaty committed:
PitchShiftEffect: formant into formant preserving

The commit changes the formant parameter Id onto the formantPreserving
name. Every dependent variable on this option is renamed
in the same way.

daschuer commented:
A conflict has developed ...

Swiftb0y committed:
Merge branch 'main' into extend_pitch_shift_effect_options

Swiftb0y commented:
My fault. fixed and pushed to this branch. @davidchocholaty, just `git pull` to get your local branch up-to-date again.

davidchocholaty commented:
> My fault. fixed and pushed to this branch. @davidchocholaty, just `git pull` to get your local branch up-to-date again.

:+1: No problem.

daschuer commented:
There is an issue in general if when an effect changes its parameters. They are not visible because of the old content in the /home/sperry/.mixxx/effects/defaults folder. This is a regression from the effect refactoring we suffer here now. Unfortunately we need to solve it else no one will see the new parameters.

daschuer commented:
After deleting the related xml file I can see the new parameters, but Mixxx is now segfaulting: 

```
Thread 66 "mixxx" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff08ff9700 (LWP 85411)]
PitchShiftEffect::processChannel (this=0x7fff98264010, pState=0x55555c283a70, 
    pInput=0x7fffbf762010, pOutput=0x7fff996b9010, engineParameters=..., 
    enableState=EffectEnableState::Enabling, groupFeatures=...)
    at ../../src/effects/backends/builtin/pitchshifteffect.cpp:126
126	    if (m_currentFormant != m_pFormantPreservingParameter->toBool()) {
(gdb) bt
#0  PitchShiftEffect::processChannel(PitchShiftGroupState*, float const*, float*, mixxx::EngineParameters const&, EffectEnableState, GroupFeatureState const&)
    (this=0x7fff98264010, pState=0x55555c283a70, pInput=0x7fffbf762010, pOutput=0x7fff996b9010, engineParameters=..., enableState=EffectEnableState::Enabling, groupFeatures=...)
    at ../../src/effects/backends/builtin/pitchshifteffect.cpp:126
#1  0x000055555594f0d3 in EngineEffect::process(ChannelHandle const&, ChannelHandle const&, float const*, float*, unsigned int, unsigned int, EffectEnableState, GroupFeatureState const&)
    (this=this@entry=0x55555c04a550, inputHandle=..., outputHandle=..., pInput=pInput@entry=0x7fffbf762010, pOutput=pOutput@entry=0x7fff996b9010, numSamples=numSamples@entry=2048, sampleRate=48000, chainEnableState=<optimized out>, groupFeatures=...) at /usr/include/c++/9/bits/unique_ptr.h:360
#2  0x0000555555953939 in EngineEffectChain::process(ChannelHandle const&, ChannelHandle const&, float*, float*, unsigned int, unsigned int, GroupFeatureState const&)
    (this=0x7fff99756010, inputHandle=..., outputHandle=..., pIn=pIn@entry=0x7fffbf762010, pOut=pOut@entry=0x7fffbf762010, numSamples=2048, sampleRate=48000, groupFeatures=...) at ../../src/engine/effects/engineeffectchain.cpp:280
#3  0x000055555595776d in EngineEffectsManager::processInner(SignalProcessingStage, ChannelHandle const&, ChannelHandle const&, float*, float*, unsigned int, unsigned int, GroupFeatureState const&, float, float)
    (this=<optimized out>, stage=<optimized out>, inputHandle=..., outputHandle=--Type <RET> for more, q to quit, c to continue without paging--
```

davidchocholaty commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -79,15 +123,23 @@ void PitchShiftEffect::processChannel(
     Q_UNUSED(enableState);
 
-    const double pitchParameter = m_pPitchParameter->value();
+    if (m_currentFormant != m_pFormantPreservingParameter->toBool()) {
+        m_currentFormant = !m_currentFormant;
```
Yeah, good idea. Anyway, didn't you mean this solution instead?

```
const bool formantPreserving = m_pFormantPreservingParameter->toBool();

if (m_currentFormant != formantPreserving) {
        m_currentFormant = formantPreserving;
```

Swiftb0y replied:
Both works. C++17 added support for `init-statement`s before the `condition`. 
https://en.cppreference.com/w/cpp/language/if
This will bind the variable scope to the if statement instead of the outer scope. I like to reduce the scope variables are bound to to be as narrow as possible.
If you find your solution easier to read/understand then go ahead and use that instead. As I said, both is legal C++

davidchocholaty replied:
> Both works. C++17 added support for `init-statement`s before the `condition`. https://en.cppreference.com/w/cpp/language/if https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r0.html 

Oh perfect, sounds cool. The new thing for me. Thank you for the links.

> This will bind the variable scope to the if statement instead of the outer scope. I like to reduce the scope variables are bound to to be as narrow as possible.

That's reasonable.

> If you find your solution easier to read/understand then go ahead and use that instead. As I said, both is legal C++

Reducing the variable scope makes sense to me.

davidchocholaty committed:
PitchShiftEffect: rename kSemitones constant

This commit renames the kSemitones constant into the kSemitonesPerOctave
name because of its more appropriate meaning.

davidchocholaty committed:
PitchShiftEffect: replace flip-flopping

This commit replaces flip-flopping approach to the boolean value
assignment with a cleaner value directly from the parameter.
The init-statement is included in the if statement before the condition
to reduce the variable scope (supported since C++17).

daschuer commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +69,48 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
     m_pPitchParameter = parameters.value("pitch");
+    m_pRangeParameter = parameters.value("range");
+    m_pSemitonesModeParameter = parameters.value("semitonesMode");
+    m_pFormantPreservingParameter = parameters.value("formant");
```
```suggestion
    m_pFormantPreservingParameter = parameters.value("formantPreserving");
```
This fixes the segfault.

davidchocholaty replied:
Oh, thank you. I'm sorry, I overlooked it.

daschuer reviewed (changes requested):
I think we need to update the version number: 

```pManifest->setVersion("2.0");```

Add a version number to the default files and discard it when there is a version mismatch.

Swiftb0y commented on src/effects/backends/builtin/pitchshifteffect.cpp:
```diff
@@ -61,12 +69,48 @@ EffectManifestPointer PitchShiftEffect::getManifest() {
 void PitchShiftEffect::loadEngineEffectParameters(
         const QMap<QString, EngineEffectParameterPointer>& parameters) {
     m_pPitchParameter = parameters.value("pitch");
+    m_pRangeParameter = parameters.value("range");
+    m_pSemitonesModeParameter = parameters.value("semitonesMode");
+    m_pFormantPreservingParameter = parameters.value("formant");
```
I think we should put these as QStringLiterals into the anonymous namespace above.

davidchocholaty replied:
Great idea, good approach to avoid the grammar mistakes which causes the segmentation fault as was found.

davidchocholaty committed:
PitchShiftEffect: fix segmentation fault

The commit fixes the segmentation fault due to a bug
in the name of the parameter value (the old id "formant" was used
instead of the new one, which is "formatPreserving").

davidchocholaty committed:
PitchShiftEffect: update version

The commit updates the version of the effect manifest from version 1.0
onto version 2.0.

davidchocholaty committed:
PitchShiftEffect: add id literals into a namespace

This commits creates literal constants in the anonymous namespace.
The constants values are the same as the manifest
parameter identifiers. The constants are created to avoid grammar bugs
due to the value has to be used twice on the other code places. Due to
the literal being used, the segmentation fault has occurred
without previous warnings or errors if the entered identifier names
for the specific parameter are not the same.

daschuer commented:
Thank you. This looks good now. We only have the pending issue that no one of our alpha users will see the new options. 
What is you plan? Does this PR fit to your future plan? 
I think this can be merged, but we need at least an issue with the 2.4 milestone to fix it.

davidchocholaty commented:
> What is you plan? Does this PR fit to your future plan?

The extension was implemented for the current Pitch shift effect implementation, anyway, it can be quite cool to test the wider range for the new implementation in [#4898](https://github.com/mixxxdj/mixxx/pull/4898). IMO the clearest way is to just merge the 'upstream/main' including the extended options into the mentioned PR. As a second, TBH merged PR looks quite better for the GSoC blog than opened, but it doesn't really matter.

> I think this can be merged, but we need at least an issue with the 2.4 milestone to fix it. Do we have one?

Yeah, of course. I have not found any issue related to this problem. If the PR will be merged, should I create the issue or do you want to do it on your own because you found the problem and maybe will better describe the situation?

Swiftb0y commented:
> Thank you. This looks good now. We only have the pending issue that no one of our alpha users will see the new options.

I may try to find a workaround during the beta. I'll open a tracking issue.

davidchocholaty commented:
> > Thank you. This looks good now. We only have the pending issue that no one of our alpha users will see the new options.
> 
> I may try to find a workaround during the beta. I'll open a tracking issue.

OK, thank you very much.

Swiftb0y commented:
Thanks

davidchocholaty commented:
Thank you.
